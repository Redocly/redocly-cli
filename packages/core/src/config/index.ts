export * from './config.js';
export * from './types.js';
export * from './rules.js';
export * from './builtIn.js';
export * from './load.js';
export * from './utils.js';
export * from './config-resolvers.js';
