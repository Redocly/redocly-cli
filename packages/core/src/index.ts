export { BundleOutputFormat, readFileFromUrl, slash } from './utils';
export { Oas3_1Types } from './types/oas3_1';
export { Oas3Types } from './types/oas3';
export { Oas2Types } from './types/oas2';
export { ConfigTypes } from './types/redocly-yaml';
export {
  Oas3Definition,
  Oas3Components,
  Oas3PathItem,
  Oas3Paths,
  Oas3ComponentName,
  Oas3Schema,
  Oas3Tag,
} from './typings/openapi';
export { Oas2Definition } from './typings/swagger';
export { StatsAccumulator, StatsName } from './typings/common';
export { normalizeTypes } from './types';
export { Stats } from './rules/other/stats';

export { Config, LintConfig, RawConfig, IGNORE_FILE } from './config/config';
export { loadConfig } from './config/load';
export { RedoclyClient } from './redocly';
export {
  Source,
  BaseResolver,
  Document,
  resolveDocument,
  ResolveError,
  YamlParseError,
  makeDocumentFromString,
} from './resolve';
export { parseYaml, stringifyYaml } from './js-yaml';
export { unescapePointer } from './ref-utils';
export { detectOpenAPI, OasMajorVersion, openAPIMajor, OasVersion } from './oas-types';
export { normalizeVisitors } from './visitors';

export {
  WalkContext,
  walkDocument,
  NormalizedProblem,
  ProblemSeverity,
  LineColLocationObject,
  LocationObject,
  Loc,
} from './walk';

export { getAstNodeByPointer, getLineColLocation } from './format/codeframes';
export { formatProblems, OutputFormat, getTotals, Totals } from './format/format';
export {
  lint,
  lint as validate,
  lintDocument,
  lintFromString,
  lintConfig,
} from './lint';
export { bundle, bundleDocument } from './bundle';
