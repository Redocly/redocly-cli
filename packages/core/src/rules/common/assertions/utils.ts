import { isRef } from '../../../ref-utils';
import { Problem, ProblemSeverity, UserContext } from '../../../walk';
import { asserts } from './asserts';

export type OrderDirection = 'asc' | 'desc';

export type OrderOptions = {
  direction: OrderDirection;
  property: string;
};

export type AssertToApply = {
  name: string;
  assertId?: string;
  conditions: any;
  message?: string;
  severity?: ProblemSeverity;
  suggest?: string[];
  runsOnKeys: boolean;
  runsOnValues: boolean;
};

export function buildVisitorObject(
  subject: string,
  context: Record<string, any>[],
  subjectVisitor: any,
) {
  if (!context) {
    return { [subject]: subjectVisitor };
  }

  let currentVisitorLevel: Record<string, any> = {};
  const visitor: Record<string, any> = currentVisitorLevel;

  for (let index = 0; index < context.length; index++) {
    const node = context[index];
    if (context.length === index + 1 && node.type === subject) {
      // Visitors don't work properly for the same type nested nodes, so
      // as a workaround for that we don't create separate visitor for the last element
      // which is the same as subject;
      // we will check includes/excludes it in the last visitor.
      continue;
    }
    const matchParentKeys = node.matchParentKeys;
    const excludeParentKeys = node.excludeParentKeys;

    if (matchParentKeys && excludeParentKeys) {
      throw new Error(
        `Both 'matchParentKeys' and 'excludeParentKeys' can't be under one context item`,
      );
    }

    if (matchParentKeys || excludeParentKeys) {
      currentVisitorLevel[node.type] = {
        skip: (_value: any, key: string) => {
          if (matchParentKeys) {
            return !matchParentKeys.includes(key);
          }
          if (excludeParentKeys) {
            return excludeParentKeys.includes(key);
          }
        },
      };
    } else {
      currentVisitorLevel[node.type] = {};
    }
    currentVisitorLevel = currentVisitorLevel[node.type];
  }

  currentVisitorLevel[subject] = subjectVisitor;

  return visitor;
}

export function buildSubjectVisitor(
  properties: string | string[],
  asserts: AssertToApply[],
  context?: Record<string, any>[],
) {
  return function (node: any, { report, location, key, type, resolve }: UserContext) {
    // We need to check context's last node if it has the same type as subject node;
    // if yes - that means we didn't create context's last node visitor,
    // so we need to handle 'matchParentKeys' and 'excludeParentKeys' conditions here;
    if (context) {
      const lastContextNode = context[context.length - 1];
      if (lastContextNode.type === type.name) {
        const matchParentKeys = lastContextNode.matchParentKeys;
        const excludeParentKeys = lastContextNode.excludeParentKeys;

        if (matchParentKeys && !matchParentKeys.includes(key)) {
          return;
        }
        if (excludeParentKeys && excludeParentKeys.includes(key)) {
          return;
        }
      }
    }

    if (properties) {
      properties = Array.isArray(properties) ? properties : [properties];
    }

    for (const assert of asserts) {
      if (properties) {
        for (const property of properties) {
          // we can have resolvable scalar so need to resolve value here.
          const value = isRef(node[property]) ? resolve(node[property])?.node : node[property];
          runAssertion(value, assert, location.child(property), report);
        }
      } else {
        runAssertion(Object.keys(node), assert, location.key(), report);
      }
    }
  };
}

export function getIntersectionLength(keys: string[], properties: string[]): number {
  const props = new Set(properties);
  let count = 0;
  for (const key of keys) {
    if (props.has(key)) {
      count++;
    }
  }
  return count;
}

export function isOrdered(value: any[], options: OrderOptions | OrderDirection): boolean {
  const direction = (options as OrderOptions).direction || (options as OrderDirection);
  const property = (options as OrderOptions).property;
  for (let i = 1; i < value.length; i++) {
    let currValue = value[i];
    let prevVal = value[i - 1];

    if (property) {
      if (!value[i][property] || !value[i - 1][property]) {
        return false; // property doesn't exist, so collection is not ordered
      }
      currValue = value[i][property];
      prevVal = value[i - 1][property];
    }

    const result = direction === 'asc' ? currValue >= prevVal : currValue <= prevVal;
    if (!result) {
      return false;
    }
  }
  return true;
}

function runAssertion(
  values: string | string[],
  assert: AssertToApply,
  location: any,
  report: (problem: Problem) => void,
) {
  const lintResult = asserts[assert.name](values, assert.conditions);
  if (!lintResult) {
    report({
      message: assert.message || `The ${assert.assertId} doesn't meet required conditions`,
      location,
      forceSeverity: assert.severity,
      suggest: assert.suggest,
      ruleId: assert.assertId,
    });
  }
}
