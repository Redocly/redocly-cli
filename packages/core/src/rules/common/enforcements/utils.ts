import { ProblemSeverity, UserContext } from '../../../walk';
import { rules as genericRules } from './generic-rules';

export type OrderDirection = 'asc' | 'desc';

export type OrderOptions = {
  direction: OrderDirection;
  property?: string;
};

export type Rule = {
  name: string;
  conditions: any;
  description: string;
  severity: ProblemSeverity;
  runsOnKeys: boolean;
  runsOnValues: boolean;
};

export const ALL_KEYS = '$keys';

/** Sets the value at path of object. If a portion of path doesn't exist, it's created.  */
export const objectSet = (path: string[], value: any) => {
  return path.reverse().reduce((acc, key) => ({[key]: acc}), value);
}

export const formRule = (lastNodeName: string, propsToRules:  {[key: string]: Rule[]}) => {
  return {
    [lastNodeName]: function(node: any, { report, location }: UserContext) {
      for (const prop of Object.keys(propsToRules)) {
        for (const rule of propsToRules[prop]) {
          const value = prop === ALL_KEYS ? node : node[prop];
          const lintResult = (genericRules as {[key: string]: any})[rule.name](value, rule.conditions);
          if (!lintResult) {
            report({
              message: rule.description,
              location: prop === ALL_KEYS ? location.key() : location.child(prop).key(),
              forceSeverity: rule.severity
            });
          }
        }
      }
    }
  }
}

export const getCounts = (node: any, properties: string[]): number => {
  let counter = 0;
  for (const prop of Object.keys(node)) {
    if (properties.includes(prop)) {
      counter++;
    }
  }
  return counter;
}

export const isOrdered = (value: any[], options: OrderOptions | OrderDirection): boolean => {
  const direction = (options as OrderOptions).direction || options as OrderDirection;
  const property = (options as OrderOptions).property;
  let result = true;
  for (let i=1; i<value.length; i++) {
    let currValue = value[i];
    let prevVal = value[i-1];

    if (property) {
      if (!value[i][property] || !value[i-1][property]) {
        return false; // property doesn't exist, so collection is not ordered
      }
      currValue = value[i][property];
      prevVal = value[i-1][property]
    }

    result = direction === 'asc' ? currValue >= prevVal : currValue <= prevVal;
    if (!result) {
      break;
    }
  }
  return result;
}