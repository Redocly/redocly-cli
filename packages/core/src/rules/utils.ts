import levenshtein = require('js-levenshtein');
import { UserContext } from '../walk';

export function oasTypeOf(value: unknown) {
  if (Array.isArray(value)) {
    return 'array';
  } else if (value === null) {
    return 'null';
  } else {
    return typeof value;
  }
}

/**
 * Checks if value matches specified JSON schema type
 *
 * @param {*} value - value to check
 * @param {JSONSchemaType} type - JSON Schema type
 * @returns boolean
 */
export function matchesJsonSchemaType(value: unknown, type: string, nullable: boolean): boolean {
  if (nullable && value === null) {
    return value === null
  }

  switch (type) {
    case 'array':
      return Array.isArray(value);
    case 'object':
      return typeof value === 'object' && value !== null && !Array.isArray(value);
    case 'null':
      return value === null;
    case 'integer':
      return Number.isInteger(value);
    default:
      return typeof value === type;
  }
}

export function missingRequiredField(type: string, field: string): string {
  return `${type} object should contain \`${field}\` field.`;
}

export function fieldNonEmpty(type: string, field: string): string {
  return `${type} object \`${field}\` must be non-empty string.`;
}

export function validateDefinedAndNonEmpty(fieldName: string, value: any, ctx: UserContext) {
  if (typeof value !== 'object') {
    return;
  }

  if (value[fieldName] === undefined) {
    ctx.report({
      message: missingRequiredField(ctx.type.name, fieldName),
      location: ctx.location.child([fieldName]).key(),
    });
  } else if (!value[fieldName]) {
    ctx.report({
      message: fieldNonEmpty(ctx.type.name, fieldName),
      location: ctx.location.child([fieldName]).key(),
    });
  }
}

export function getSuggest(given: string, variants: string[]): string[] {
  if (typeof given !== 'string' || !variants.length) return [];

  const distances: Array<{ variant: string; distance: number }> = [];

  for (let i = 0; i < variants.length; i++) {
    const distance = levenshtein(given, variants[i]);
    if (distance < 4) {
      distances.push({ distance, variant: variants[i] });
    }
  }

  distances.sort((a, b) => a.distance - b.distance);

  // if (bestMatch.distance <= 4) return bestMatch.string;
  return distances.map((d) => d.variant);
}