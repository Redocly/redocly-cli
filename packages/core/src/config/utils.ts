import { green, yellow } from 'colorette';
import { assignExisting } from '../utils';
import { Config } from './config';

import type {
  Api,
  DeprecatedRawConfig,
  Plugin,
  RawConfig,
  RawResolveConfig,
  ResolveConfig,
  ResolvedLintConfig,
  RulesFields,
} from './types';

export function parsePresetName(presetName: string): { pluginId: string; configName: string } {
  if (presetName.indexOf('/') > -1) {
    const [pluginId, configName] = presetName.split('/');
    return { pluginId, configName };
  } else {
    return { pluginId: '', configName: presetName };
  }
}

export function transformApiDefinitionsToApis(
  apiDefinitions: Record<string, string> = {},
): Record<string, Api> {
  let apis: Record<string, Api> = {};
  for (const [apiName, apiPath] of Object.entries(apiDefinitions)) {
    apis[apiName] = { root: apiPath };
  }
  return apis;
}

export function prefixRules<T extends Record<string, any>>(rules: T, prefix: string) {
  if (!prefix) return rules;

  const res: any = {};
  for (const name of Object.keys(rules)) {
    res[`${prefix}/${name}`] = rules[name];
  }

  return res;
}

export function mergeExtends(rulesConfList: ResolvedLintConfig[]) {
  const result: Omit<ResolvedLintConfig, RulesFields> &
    Required<Pick<ResolvedLintConfig, RulesFields>> = {
    rules: {},
    oas2Rules: {},
    oas3_0Rules: {},
    oas3_1Rules: {},

    preprocessors: {},
    oas2Preprocessors: {},
    oas3_0Preprocessors: {},
    oas3_1Preprocessors: {},

    decorators: {},
    oas2Decorators: {},
    oas3_0Decorators: {},
    oas3_1Decorators: {},

    plugins: [],
    pluginPaths: [],
    extendPaths: [],
  };

  for (let rulesConf of rulesConfList) {
    if (rulesConf.extends) {
      throw new Error(
        `\`extends\` is not supported in shared configs yet: ${JSON.stringify(
          rulesConf,
          null,
          2,
        )}.`,
      );
    }

    Object.assign(result.rules, rulesConf.rules);
    Object.assign(result.oas2Rules, rulesConf.oas2Rules);
    assignExisting(result.oas2Rules, rulesConf.rules || {});
    Object.assign(result.oas3_0Rules, rulesConf.oas3_0Rules);
    assignExisting(result.oas3_0Rules, rulesConf.rules || {});
    Object.assign(result.oas3_1Rules, rulesConf.oas3_1Rules);
    assignExisting(result.oas3_1Rules, rulesConf.rules || {});

    Object.assign(result.preprocessors, rulesConf.preprocessors);
    Object.assign(result.oas2Preprocessors, rulesConf.oas2Preprocessors);
    assignExisting(result.oas2Preprocessors, rulesConf.preprocessors || {});
    Object.assign(result.oas3_0Preprocessors, rulesConf.oas3_0Preprocessors);
    assignExisting(result.oas3_0Preprocessors, rulesConf.preprocessors || {});
    Object.assign(result.oas3_1Preprocessors, rulesConf.oas3_1Preprocessors);
    assignExisting(result.oas3_1Preprocessors, rulesConf.preprocessors || {});

    Object.assign(result.decorators, rulesConf.decorators);
    Object.assign(result.oas2Decorators, rulesConf.oas2Decorators);
    assignExisting(result.oas2Decorators, rulesConf.decorators || {});
    Object.assign(result.oas3_0Decorators, rulesConf.oas3_0Decorators);
    assignExisting(result.oas3_0Decorators, rulesConf.decorators || {});
    Object.assign(result.oas3_1Decorators, rulesConf.oas3_1Decorators);
    assignExisting(result.oas3_1Decorators, rulesConf.decorators || {});

    result.plugins!.push(...(rulesConf.plugins || []));
    result.pluginPaths!.push(...(rulesConf.pluginPaths || []));
    result.extendPaths!.push(...new Set(rulesConf.extendPaths));
  }

  return result;
}

export function getMergedConfig(config: Config, entrypointAlias?: string): Config {
  const extendPaths = [
    ...Object.values(config.apis).map((api) => api?.lint?.extendPaths),
    config.rawConfig?.lint?.extendPaths,
  ]
    .flat()
    .filter(Boolean) as string[];

  const pluginPaths = [
    ...Object.values(config.apis).map((api) => api?.lint?.pluginPaths),
    config.rawConfig?.lint?.pluginPaths,
  ]
    .flat()
    .filter(Boolean) as string[];

  return entrypointAlias
    ? new Config(
        {
          ...config.rawConfig,
          lint: {
            ...(config.apis[entrypointAlias]
              ? config.apis[entrypointAlias].lint
              : config.rawConfig.lint),
            extendPaths,
            pluginPaths,
          },
          'features.openapi': {
            ...config['features.openapi'],
            ...config.apis[entrypointAlias]?.['features.openapi'],
          },
          'features.mockServer': {
            ...config['features.mockServer'],
            ...config.apis[entrypointAlias]?.['features.mockServer'],
          },
          // TODO: merge everything else here
        },
        config.configFile,
      )
    : config;
}

export function transformConfig(rawConfig: DeprecatedRawConfig | RawConfig): RawConfig {
  if ((rawConfig as RawConfig).apis && (rawConfig as DeprecatedRawConfig).apiDefinitions) {
    throw new Error("Do not use 'apiDefinitions' field. Use 'apis' instead.\n");
  }
  if (
    (rawConfig as RawConfig)['features.openapi'] &&
    (rawConfig as DeprecatedRawConfig).referenceDocs
  ) {
    throw new Error("Do not use 'referenceDocs' field. Use 'features.openapi' instead.\n");
  }
  const { apiDefinitions, referenceDocs, ...rest } = rawConfig as DeprecatedRawConfig & RawConfig;
  if (apiDefinitions) {
    process.stderr.write(
      `The ${yellow('apiDefinitions')} field is deprecated. Use ${green(
        'apis',
      )} instead. Read more about this change: https://redocly.com/docs/api-registry/guides/migration-guide-config-file/#changed-properties\n`,
    );
  }
  if (referenceDocs) {
    process.stderr.write(
      `The ${yellow('referenceDocs')} field is deprecated. Use ${green(
        'features.openapi',
      )} instead. Read more about this change: https://redocly.com/docs/api-registry/guides/migration-guide-config-file/#changed-properties\n`,
    );
  }
  return {
    'features.openapi': referenceDocs,
    apis: transformApiDefinitionsToApis(apiDefinitions),
    ...rest,
  };
}

export function getResolveConfig(resolve?: RawResolveConfig): ResolveConfig {
  return {
    http: {
      headers: resolve?.http?.headers ?? [],
      customFetch: undefined,
    },
  };
}

export function getUniquePlugins(plugins: Plugin[]): Plugin[] {
  const seen = new Set();
  const results = [];
  for (const p of plugins) {
    if (!seen.has(p.id)) {
      results.push(p);
      seen.add(p.id);
    } else if (p.id) {
      process.stderr.write(
        `Duplicate plugin id "${yellow(p.id)}".\n`,
      );
    }
  }
  return results;
}
