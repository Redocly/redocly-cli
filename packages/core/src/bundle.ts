import isEqual = require('lodash.isequal');
import { BaseResolver, resolveDocument, Document } from './resolve';
import { Oas3Rule, normalizeVisitors, Oas3Visitor, Oas2Visitor } from './visitors';
import { Oas3Types } from './types/oas3';
import { Oas2Types } from './types/oas2';
import { Oas3_1Types } from './types/oas3_1';
import { NormalizedNodeType, normalizeTypes, NodeType } from './types';
import { WalkContext, walkDocument, UserContext, ResolveResult } from './walk';
import { detectOpenAPI, openAPIMajor, OasMajorVersion } from './oas-types';
import { Location, refBaseName } from './ref-utils';
import type { Config, LintConfig } from './config/config';
import { initRules } from './config/rules';
import { reportUnresolvedRef } from './rules/no-unresolved-refs';
import { isPlainObject } from './utils';
import { OasRef } from './typings/openapi';

export type Oas3RuleSet = Record<string, Oas3Rule>;

export enum OasVersion {
  Version2 = 'oas2',
  Version3_0 = 'oas3_0',
  Version3_1 = 'oas3_1',
}

export async function bundle(opts: {
  ref?: string;
  doc?: Document;
  externalRefResolver?: BaseResolver;
  config: Config;
  dereference?: boolean;
  base?: string;
}) {
  const {
    ref,
    doc,
    externalRefResolver = new BaseResolver(opts.config.resolve),
    base = null,
  } = opts;
  if (!(ref || doc)) {
    throw new Error('Document or reference is required.\n');
  }

  const document = doc !== undefined ? doc : await externalRefResolver.resolveDocument(base, ref!, true);

  if (document instanceof Error) {
    throw document;
  }

  return bundleDocument({
    document,
    ...opts,
    config: opts.config.lint,
    externalRefResolver,
  });
}

type BundleContext = WalkContext;

export async function bundleDocument(opts: {
  document: Document;
  config: LintConfig;
  customTypes?: Record<string, NodeType>;
  externalRefResolver: BaseResolver;
  dereference?: boolean;
}) {
  const { document, config, customTypes, externalRefResolver, dereference = false } = opts;
  const oasVersion = detectOpenAPI(document.parsed);
  const oasMajorVersion = openAPIMajor(oasVersion);
  const rules = config.getRulesForOasVersion(oasMajorVersion);
  const types = normalizeTypes(
    config.extendTypes(
      customTypes ?? oasMajorVersion === OasMajorVersion.Version3 ? (oasVersion === OasVersion.Version3_1 ? Oas3_1Types : Oas3Types) : Oas2Types,
      oasVersion,
    ),
    config,
  );

  const preprocessors = initRules(rules as any, config, 'preprocessors', oasVersion);
  const decorators = initRules(rules as any, config, 'decorators', oasVersion);
  const ctx: BundleContext = {
    problems: [],
    oasVersion: oasVersion,
    refTypes: new Map<string, NormalizedNodeType>(),
  };

  const bundleVisitor = normalizeVisitors(
    [
      ...preprocessors,
      {
        severity: 'error',
        ruleId: 'bundler',
        visitor: makeBundleVisitor(oasMajorVersion, dereference, document),
      },
      ...decorators,
    ],
    types,
  );

  const resolvedRefMap = await resolveDocument({
    rootDocument: document,
    rootType: types.DefinitionRoot,
    externalRefResolver,
  });

  walkDocument({
    document,
    rootType: types.DefinitionRoot as NormalizedNodeType,
    normalizedVisitors: bundleVisitor,
    resolvedRefMap,
    ctx,
  });

  return {
    bundle: document,
    problems: ctx.problems.map((problem) => config.addProblemToIgnore(problem)),
    fileDependencies: externalRefResolver.getFiles(),
    rootType: types.DefinitionRoot,
    refTypes: ctx.refTypes,
  };
}

function mapTypeToComponent(typeName: string, version: OasMajorVersion) {
  switch (version) {
    case OasMajorVersion.Version3:
      switch (typeName) {
        case 'Schema':
          return 'schemas';
        case 'Parameter':
          return 'parameters';
        case 'Response':
          return 'responses';
        case 'Example':
          return 'examples';
        case 'RequestBody':
          return 'requestBodies';
        case 'Header':
          return 'headers';
        case 'SecuritySchema':
          return 'securitySchemes';
        case 'Link':
          return 'links';
        case 'Callback':
          return 'callbacks';
        default:
          return null;
      }
    case OasMajorVersion.Version2:
      switch (typeName) {
        case 'Schema':
          return 'definitions';
        case 'Parameter':
          return 'parameters';
        case 'Response':
          return 'responses';
        default:
          return null;
      }
  }
}

// function oas3Move

function makeBundleVisitor(version: OasMajorVersion, dereference: boolean, rootDocument: Document) {
  let components: Record<string, Record<string, any>>;

  const visitor: Oas3Visitor | Oas2Visitor = {
    ref: {
      leave(node, ctx, resolved) {
        if (!resolved.location || resolved.node === undefined) {
          reportUnresolvedRef(resolved, ctx.report, ctx.location);
          return;
        }
        if (
          resolved.location.source === rootDocument.source &&
          resolved.location.source === ctx.location.source &&
          ctx.type.name !== 'scalar' &&
          !dereference
        ) {
          return;
        }
        const componentType = mapTypeToComponent(ctx.type.name, version);
        if (!componentType) {
          replaceRef(node, resolved, ctx);
        } else {
          if (dereference) {
            saveComponent(componentType, resolved, ctx);
            replaceRef(node, resolved, ctx);
          } else {
            node.$ref = saveComponent(componentType, resolved, ctx);
          }
        }
      },
    },
    DefinitionRoot: {
      enter(root: any) {
        if (version === OasMajorVersion.Version3) {
          components = root.components = root.components || {};
        } else if (version === OasMajorVersion.Version2) {
          components = root;
        }
      },
    },
  };

  if (version === OasMajorVersion.Version3) {
    visitor.DiscriminatorMapping = {
      leave(mapping: Record<string, string>, ctx: any) {
        for (const name of Object.keys(mapping)) {
          const $ref = mapping[name];
          const resolved = ctx.resolve({ $ref });
          if (!resolved.location || resolved.node === undefined) {
            reportUnresolvedRef(resolved, ctx.report, ctx.location.child(name));
            return;
          }

          const componentType = mapTypeToComponent('Schema', version)!;
          if (dereference) {
            saveComponent(componentType, resolved, ctx);
          } else {
            mapping[name] = saveComponent(componentType, resolved, ctx);
          }
        }
      },
    };
  }

  function replaceRef(ref: OasRef, resolved: ResolveResult<any>, ctx: UserContext) {
    if (!isPlainObject(resolved.node)) {
      ctx.parent[ctx.key] = resolved.node;
    } else {
      // @ts-ignore
      delete ref.$ref;
      Object.assign(ref, resolved.node);
    }
  }

  function saveComponent(
    componentType: string,
    target: { node: any; location: Location },
    ctx: UserContext,
  ) {
    components[componentType] = components[componentType] || {};
    const name = getComponentName(target, componentType, ctx);
    components[componentType][name] = target.node;
    if (version === OasMajorVersion.Version3) {
      return `#/components/${componentType}/${name}`;
    } else {
      return `#/${componentType}/${name}`;
    }
  }

  function getComponentName(
    target: { node: any; location: Location },
    componentType: string,
    ctx: UserContext,
  ) {
    const [fileRef, pointer] = [target.location.source.absoluteRef, target.location.pointer];
    const componentsGroup = components[componentType];

    let name = '';

    const refParts = pointer.slice(2).split('/').filter(Boolean); // slice(2) removes "#/"
    while (refParts.length > 0) {
      name = refParts.pop() + (name ? `-${name}` : '');
      if (
        !componentsGroup ||
        !componentsGroup[name] ||
        isEqual(componentsGroup[name], target.node)
      ) {
        return name;
      }
    }

    name = refBaseName(fileRef) + (name ? `_${name}` : '');
    if (!componentsGroup[name] || isEqual(componentsGroup[name], target.node)) {
      return name;
    }

    const prevName = name;
    let serialId = 2;
    while (componentsGroup[name] && !isEqual(componentsGroup[name], target.node)) {
      name = `${prevName}-${serialId}`;
      serialId++;
    }

    if (!componentsGroup[name]) {
      ctx.report({
        message: `Two schemas are referenced with the same name but different content. Renamed ${prevName} to ${name}.`,
        location: ctx.location,
        forceSeverity: 'warn',
      });
    }

    return name;
  }

  return visitor;
}
