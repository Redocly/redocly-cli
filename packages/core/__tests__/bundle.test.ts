import outdent from 'outdent';
import * as path from 'path';

import { bundleDocument, bundle } from '../src/bundle';

import { parseYamlToDocument, yamlSerializer } from './utils';
import { LintConfig, Config } from '../src/config/config';
import { BaseResolver } from '../src/resolve';

describe('bundle', () => {
  expect.addSnapshotSerializer(yamlSerializer);

  const testDocument = parseYamlToDocument(
    outdent`
      openapi: 3.0.0
      paths:
        /pet:
          get:
            operationId: get
            parameters:
              - $ref: '#/components/parameters/shared_a'
              - name: get_b
          post:
            operationId: post
            parameters:
              - $ref: '#/components/parameters/shared_a'
      components:
        parameters:
          shared_a:
            name: shared-a
    `,
    '',
  );

  it('change nothing with only internal refs', async () => {
    const { bundle, problems } = await bundleDocument({
      document: testDocument,
      externalRefResolver: new BaseResolver(),
      config: new LintConfig({}),
    });

    const origCopy = JSON.parse(JSON.stringify(testDocument.parsed));

    expect(problems).toHaveLength(0);
    expect(bundle.parsed).toEqual(origCopy);
  });

  it('should bundle external refs', async () => {
    const { bundle: res, problems } = await bundle({
      config: new Config({}),
      ref: path.join(__dirname, 'fixtures/refs/openapi-with-external-refs.yaml'),
    });
    expect(problems).toHaveLength(0);
    expect(res.parsed).toMatchSnapshot();
  });

  it('should bundle external refs and warn for conflicting names', async () => {
    const { bundle: res, problems } = await bundle({
      config: new Config({}),
      ref: path.join(__dirname, 'fixtures/refs/openapi-with-external-refs-conflicting-names.yaml'),
    });
    expect(problems).toHaveLength(1);
    expect(problems[0].message).toEqual(
      `Two schemas are referenced with the same name but different content. Renamed param-b to param-b-2.`,
    );
    expect(res.parsed).toMatchSnapshot();
  });

  it('should dereferenced correctly when used with dereference', async () => {
    const { bundle: res, problems } = await bundleDocument({
      externalRefResolver: new BaseResolver(),
      config: new LintConfig({}),
      document: testDocument,
      dereference: true,
    });

    expect(problems).toHaveLength(0);
    expect(res.parsed).toMatchSnapshot();
  });

  it('should place referenced schema inline when referenced schema name resolves to original schema name', async () => {
    const { bundle: res, problems } = await bundle({
      config: new Config({}),
      ref: path.join(__dirname, 'fixtures/refs/externalref.yaml'),
    });

    expect(problems).toHaveLength(0);
    expect(res.parsed).toMatchSnapshot();
  });

  it('should not place referened schema inline when component in question is not of type "schemas"', async () => {
    const { bundle: res, problems } = await bundle({
      config: new Config({}),
      ref: path.join(__dirname, 'fixtures/refs/external-request-body.yaml'),
    });

    expect(problems).toHaveLength(0);
    expect(res.parsed).toMatchSnapshot();
  });

  it('should pull hosted schema', async () => {
    const fetchMock = jest.fn(
      () => Promise.resolve({
        ok: true,
        text: () => 'External schema content',
        headers: {
          get: () => ''
        }
      })
    );

    const { bundle: res, problems } = await bundle({
      config: new Config({}),
      externalRefResolver: new BaseResolver({
        http: {
          customFetch: fetchMock,
          headers: []
        }
      }),
      ref: path.join(__dirname, 'fixtures/refs/hosted.yaml')
    });

    expect(problems).toHaveLength(0);
    expect(fetchMock).toHaveBeenCalledWith(
      "https://someexternal.schema",
      {
        headers: {}
      }
    );
    expect(res.parsed).toMatchSnapshot();
  })
});
