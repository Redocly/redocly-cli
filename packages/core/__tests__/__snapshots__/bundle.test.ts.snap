// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bundle should bundle external refs 1`] = `
openapi: 3.0.0
paths:
  /pet:
    parameters:
      - $ref: '#/components/parameters/path-param'
    put:
      parameters:
        - $ref: '#/components/parameters/param-a'
        - $ref: '#/components/parameters/param-b'
        - name: test
    get:
      parameters:
        - $ref: '#/components/parameters/param-a'
        - $ref: '#/components/parameters/param-c'
    post:
      parameters:
        - $ref: '#/paths/~1pet/put/parameters/2'
components:
  parameters:
    param-a:
      name: param_a
      schema:
        $ref: '#/components/schemas/schema-a'
      examples:
        first:
          $ref: '#/components/examples/param-a-first'
        second:
          $ref: '#/components/examples/second'
    path-param:
      name: path_param
      examples:
        first:
          $ref: '#/components/examples/first'
    param-c:
      name: param_c
    param-b:
      name: param_b
  examples:
    first:
      value: b1
    param-a-first:
      value: a1
    second:
      value: a2
  schemas:
    schema-a:
      type: string

`;

exports[`bundle should bundle external refs and warn for conflicting names 1`] = `
openapi: 3.0.0
paths:
  /pet:
    parameters:
      - $ref: '#/components/parameters/path-param'
    put:
      parameters:
        - $ref: '#/components/parameters/param-b'
        - $ref: '#/components/parameters/param-b-2'
    get:
      parameters:
        - $ref: '#/components/parameters/param-a'
        - $ref: '#/components/parameters/param-c'
components:
  parameters:
    param-a:
      name: test
    param-b:
      name: param_b_alt
    path-param:
      name: path_param
    param-c:
      name: param_c
    param-b-2:
      name: param_b

`;

exports[`bundle should dereferenced correctly when used with dereference 1`] = `
openapi: 3.0.0
paths:
  /pet:
    get:
      operationId: get
      parameters:
        - name: shared-a
        - name: get_b
    post:
      operationId: post
      parameters:
        - name: shared-a
components:
  parameters:
    shared_a:
      name: shared-a

`;

exports[`bundle should not place referened schema inline when component in question is not of type "schemas" 1`] = `
openapi: 3.0.0
paths:
  /pet:
    post:
      requestBody:
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/requestBody'
components:
  schemas:
    requestBody:
      content:
        application/json:
          schema:
            type: object
            properties:
              a:
                type: string
              b:
                type: number

`;

exports[`bundle should place referenced schema inline when referenced schema name resolves to original schema name 1`] = `
openapi: 3.1.0
info:
  title: My API
  description: It ain't so wonderful, but at least it's mine.
  version: '1.0'
  contact:
    email: me@theintenet.com
    name: me
paths:
  /test:
    get:
      summary: test
      responses:
        '200':
          description: test
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/vendor'
components:
  schemas:
    vendor:
      title: vendor
      type: object
      description: Vendors
      properties:
        key:
          type: string
          description: System-assigned key for the vendor.
          readOnly: true
        id:
          type: string
          description: >
            Unique identifier of the vendor.

            You must specify a unique vendor ID when creating a vendor unless
            document sequencing is configured, in which case the ID is
            auto-generated.
        name:
          type: string
          description: Name of the vendor.
        isOneTimeUse:
          type: boolean
          description: One-time use
          default: false
    myvendor:
      $ref: '#/components/schemas/vendor'
    simple:
      type: string
    A:
      type: string
    test:
      $ref: '#/components/schemas/rename-2'
    rename:
      type: string
    rename-2:
      type: number

`;

exports[`bundle should pull hosted schema 1`] = `
openapi: 3.0.3
info:
  title: bugtest
  version: '1.0'
  description: Demo
  license:
    name: DEMO
    url: https://demo.com
servers:
  - url: http://demo.com/api
paths:
  /customer:
    summary: Customer scope
    get:
      summary: Get demo no refs
      operationId: GetCustomer
      description: Returns Demo No Refs
      responses:
        '200':
          description: Demo No Refs
          content:
            application/json:
              schema:
                $ref: '#/components/schemas/Customer'
components:
  schemas:
    Customer:
      type: object
      properties:
        customerName:
          type: string
        accounts:
          type: array
          items:
            $ref: '#/components/schemas/someexternal'
    someexternal: External schema content

`;
